### PR TITLE
Logging: Implement "search within log files"

### LINKED ISSUE #41257: Logging: Add a "search within files" field and a related "search results" view
Splitting this off from https://github.com/woocommerce/woocommerce/issues/40644:

> Add a search function that does a string search within the current, possibly filtered set of log files. This would introduce a "search results" view in a list table-style format, with each row being a matched line from a log file. Clicking a result would take you to the view of that log file, directly to that particular line.

* Add a search field to the top of the log files list table view. The search would be applied to the current set of log files, meaning you could filter by source (and possibly by date at some point) before running the search. It would ignore pagination though.
* Default would be a plain string partial match search, but it might be worth it to let people write regex in the search field too.
* Submitting the search would iterate through each file in the filtered set, stream that file, and then check for matches on each line of the file. Not sure about the speed/performance implications of this yet, but if we stream each file one at a time, we hopefully won't run into memory issues.
* Matched lines would be stored in an array. Each line would then be parsed, with the match part highlighted, and output to a "Search Results" view.
* Each search result would be linked to the line number in the corresponding file.
* Should we cache the search results for the last search so the user can return to the search results screen without having to re-run the search?

### PR DESCRIPTION
**Pull request by coreymckrill**
### Changes proposed in this Pull Request:

Adds a search field to the log files list table screen that will do a partial match string search on the contents of all the files currently shown in the list table (and subsequent pages, if there is pagination). Introduces a new Search Results view that shows a list table of the matches resulting from a search. Each match shows the entire line from a log file that contains the match, and the matched string itself is highlighted. Each match also includes a link that goes to that particular line in situ in the corresponding log file.

Fixes #41257

Screen Shot 2023-11-16 at 16 39 26

#### Setup

1. After checking out this branch, you may need to update the autoloader. From the **plugins/woocommerce** directory, run `composer dump-autoload`. You also will probably need to rebuild the stylesheet for the Logs screen. For this, run `pnpm --filter=woocommerce/client/legacy run build`.
1. In your test site's **wp-config.php** file, add this line: `define( 'WC_LOG_HANDLER', 'Automattic\\WooCommerce\\Internal\\Admin\\Logging\\LogHandlerFileV2' );`. This is what allows you to see the new log file views.
2. Install [this plugin](https://gist.github.com/coreymckrill/2968911c8b63c556e7787f2cdc65f966) on your test site for generating test log files. Probably easiest to copy the file into the mu-plugins directory.
3. Open two browser tabs: the Logs screen (WP Admin > WooCommerce > Status > Logs) and the WooCommerce Tools screen (WP Admin > WooCommerce > Status > Tools). On the Tools screen you should see the **Debug Log Generator** tool at the top of the list.

#### Tests

1. On the Logs screen, you should see a list table, which may already have some log files listed in it. Even if there are some, go ahead and generate some additional logs on the Tools screen so that we have enough different log files and sources.
2. On the Logs screen, type something in the field labeled "Search within these files". E.g. try `debug`. Then click Search.
3. This should take you to a "Search results" view. You should have a few results since the test plugin for generating log files adds some entries with "DEBUG" level severity. Note that the string match is not case sensitive, and the "DEBUG" substring in each search result is highlighted in yellow.
4. Click on the linked line number for one of the results. This should take you to the single file view for the corresponding log file, and the line with the matched substring should be highlighted.
5. While viewing this single file, click "Delete permanently" to delete the log file. After confirming, this should take you back to the list table of all the available log files. Now try searching for exactly the same string again.
6. On the search results screen, the matches that were in the file that you just deleted should no longer appear.
7. While still on the search results screen, click the Screen Options tab at the top and change the number of items per page to something low, like `2`, so you can test the pagination of search results.

### Changelog entry

-   [ ] Automatically create a changelog entry from the details below.

<details>

#### Significance

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Comment 

</details>

### CODE REVIEW

File: plugins/woocommerce/src/Internal/Admin/Logging/PageController.php [hunk 1]
```diff
@@ -366,17 +410,19 @@ private function get_list_table(): ListTable {
...
+			// Ensure list table columns are initialized early enough to enable column hiding, if available.
+			$list_table->prepare_column_headers();
 
 			add_screen_option(
 				'per_page',
 				array(
-					'default' => 20,
-					'option'  => ListTable::PER_PAGE_USER_OPTION_KEY,
+					'default' => $list_table->get_per_page_default(),
+					'option'  => get_class( $list_table )::PER_PAGE_USER_OPTION_KEY,
```
Comment (jorgeatorres) [inline plugins/woocommerce/src/Internal/Admin/Logging/PageController.php]: I believe the `get_class()` here doesn't add much and probably reduces readability. You can access the constants directly with `$this->file_controller::SEARCH_MAX_RESULTS`, as you already do in different places.

File: plugins/woocommerce/src/Internal/Admin/Logging/PageController.php [hunk 2]
```diff
@@ -523,8 +561,65 @@ private function format_line( string $text, int $line_number ): string {
...
+							id="logs-search-field"
+							class="wc-logs-search-field"
+							type="text"
+							name="search"
+							value="<?php echo esc_attr( $params['search'] ); ?>"
+						/>
+					</label>
+					<?php submit_button( __( 'Search', 'woocommerce' ), 'secondary', null, false ); ?>
+				</fieldset>
+				<?php if ( $file_count >= get_class( $this->file_controller )::SEARCH_MAX_FILES ) : ?>
```
Comment (jorgeatorres) [inline plugins/woocommerce/src/Internal/Admin/Logging/PageController.php]: `get_class()`.

File: plugins/woocommerce/src/Internal/Admin/Logging/PageController.php [hunk 3]
```diff
@@ -523,8 +561,65 @@ private function format_line( string $text, int $line_number ): string {
...
+					<div class="wc-logs-search-notice">
+						<?php
+						printf(
+							// translators: %s is a number.
+							esc_html__(
+								'⚠️ Only %s files can be searched at one time. Try filtering the file list before searching.',
+								'woocommerce'
+							),
+							esc_html( number_format_i18n( get_class( $this->file_controller )::SEARCH_MAX_FILES ) )
```
Comment (jorgeatorres) [inline plugins/woocommerce/src/Internal/Admin/Logging/PageController.php]: `get_class()`.

File: plugins/woocommerce/src/Internal/Admin/Logging/FileV2/SearchListTable.php [hunk 4]
```diff
@@ -0,0 +1,227 @@
...
+		if ( is_wp_error( $total_items ) ) {
+			printf(
+				'<div class="notice notice-warning"><p>%s</p></div>',
+				esc_html( $total_items->get_error_message() )
+			);
+
+			return;
+		}
+
+		if ( $total_items > get_class( $this->file_controller )::SEARCH_MAX_RESULTS ) {
```
Comment (jorgeatorres) [inline plugins/woocommerce/src/Internal/Admin/Logging/FileV2/SearchListTable.php]: `get_class()`.

File: plugins/woocommerce/src/Internal/Admin/Logging/FileV2/SearchListTable.php [hunk 5]
```diff
@@ -0,0 +1,227 @@
...
+			printf(
+				'<div class="notice notice-info"><p>%s</p></div>',
+				sprintf(
+					// translators: %s is a number.
+					esc_html__( 'The number of search results has exceeded the limit of %s. Try narrowing your search.', 'woocommerce' ),
+					esc_html( number_format_i18n( get_class( $this->file_controller )::SEARCH_MAX_RESULTS ) )
```
Comment (jorgeatorres) [inline plugins/woocommerce/src/Internal/Admin/Logging/FileV2/SearchListTable.php]: `get_class()`.

File: plugins/woocommerce/src/Internal/Admin/Logging/FileV2/FileController.php [hunk 6]
```diff
@@ -299,9 +338,113 @@ public function delete_files( array $file_ids ): int {
...
+					if ( false !== stripos( $sanitized_line, $search ) ) {
+						$matched_lines[] = array(
+							'file_id'     => $file->get_file_id(),
+							'line_number' => $line_number,
+							'line'        => $sanitized_line,
+						);
+					}
+
+					if ( count( $matched_lines ) > self::SEARCH_MAX_RESULTS ) {
+						break 2;
```
Comment (jorgeatorres) [inline plugins/woocommerce/src/Internal/Admin/Logging/FileV2/FileController.php:415]: When we `break 2` here, the `$file->close_stream()` below doesn't get executed, which isn't truly a problem, but might be worth adding that just to keep things tidy.

**Review by jorgeatorres — CHANGES_REQUESTED** [review]
Hi @coreymckrill!

Good job on this one and the new logging implementation in general. Things are looking really good!
I left some comments, though nothing too serious.

I did have a question, though not something that we necessarily need to address in this PR (or even at all). When searching, my gut says that most of the time you'd like to see most recent results first instead of first to last. Would you consider filing an enhancement to make that either configurable or default behavior in the future?

I guess it can be tricky to read files bottom to top in a memory efficient way, and you can always refine the search terms to obtain better results, but thought I'd mention it anyways.

File: plugins/woocommerce/src/Internal/Admin/Logging/PageController.php (hunk 1 above)
Comment (coreymckrill) [reply plugins/woocommerce/src/Internal/Admin/Logging/PageController.php]: Good call. I feel like I tried it the more direct way, but my IDE protested, so I just used `get_class`. My IDE seems fine with it now, though, so updated: 99e2c0e562ad803dd8477693698d024cb41d5680

File: plugins/woocommerce/src/Internal/Admin/Logging/PageController.php (hunk 2 above)
Comment (coreymckrill) [reply plugins/woocommerce/src/Internal/Admin/Logging/PageController.php]: 99e2c0e562ad803dd8477693698d024cb41d5680

File: plugins/woocommerce/src/Internal/Admin/Logging/PageController.php (hunk 3 above)
Comment (coreymckrill) [reply plugins/woocommerce/src/Internal/Admin/Logging/PageController.php]: 99e2c0e562ad803dd8477693698d024cb41d5680

File: plugins/woocommerce/src/Internal/Admin/Logging/FileV2/SearchListTable.php (hunk 4 above)
Comment (coreymckrill) [reply plugins/woocommerce/src/Internal/Admin/Logging/FileV2/SearchListTable.php]: 99e2c0e562ad803dd8477693698d024cb41d5680

File: plugins/woocommerce/src/Internal/Admin/Logging/FileV2/SearchListTable.php (hunk 5 above)
Comment (coreymckrill) [reply plugins/woocommerce/src/Internal/Admin/Logging/FileV2/SearchListTable.php]: 99e2c0e562ad803dd8477693698d024cb41d5680

File: plugins/woocommerce/src/Internal/Admin/Logging/FileV2/FileController.php (hunk 6 above)
Comment (coreymckrill) [reply plugins/woocommerce/src/Internal/Admin/Logging/FileV2/FileController.php:415]: Yeah, the File class also does a `close_stream` in the `__destruct` magic method, but you're right that it's worth keeping things tidy: 280022cb7518b110f44b464eb091c7fa51ff4d6b

**Comment by coreymckrill** [comment]
@jorgeatorres Thanks for the review!

> When searching, my gut says that most of the time you'd like to see most recent results first instead of first to last. Would you consider filing an enhancement to make that either configurable or default behavior in the future?

You could be right about this, although I can definitely see the argument for the "earliest to latest" order as well. Along the lines of what you mention, I suspect that reading the files from bottom to top would be tricky, since as far as I know none of the native PHP filesystem functions really support this.

One thing we might be able to do, in a later PR overhauling the log handler class itself (I've already started on it!), is have a configurable option so that the files themselves could be written by prepending new lines to the top rather than appending them to the bottom. That way, these search results would naturally be ordered with the most recent results first.

**Review by jorgeatorres — APPROVED** [review]
LGTM 💯

**Pull request merged**